### PR TITLE
feat: Temperature dependent single phase thermal conductivity

### PR DESCRIPTION
**Pull request by sytuannguyen**
This PR is part of the project for non-linear thermal constitutive behavior from TotalEnergies. It tackles the problem of linear temperature dependent single phase effective thermal conductivity of porous medium.

thermalConductivity = defaultThermalConductivity + dThermalConductivity_dT * (T - T_reference)

where defaultThermalConductivity, dThermalConductivity_dT and T_reference are given in the xml inputs. The existing linear model in GEOS corresponds to the particular case of a constant thermal conductivity, i.e. dThermalConductivity_dT = 0.

- [x] Implement the non linear thermal constitutive model
- [x] Update the thermal single phase flow with respect to non linear thermal behavior
- [x] Update xml files to add new inputs: dThermalConductivity_dT and T_reference
- [x] Integrated test
- [x] Unit test
- [x] Validation example: doc and Python script for validation

### CODE REVIEW

**Review by herve-gross — APPROVED** [review]
Thank you Sy Tuan. Lots of change in this PR, but everything looks in order.

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (jhuang2601) [inline src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp]: This function is sort of redundant and will be overridden by `updateFromTemperature` and can be removed?

File: src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp
Comment (jhuang2601) [inline src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp]:
```suggestion
  arrayView1d< real64 const > const & temperature = subRegion.template getField< fields::flow::temperature >();
  conductivityMaterial.updateFromTemperature( temperature );
```

File: src/coreComponents/physicsSolvers/fluidFlow/ThermalSinglePhaseFVMKernels.hpp
Comment (jhuang2601) [inline src/coreComponents/physicsSolvers/fluidFlow/ThermalSinglePhaseFVMKernels.hpp:361]:
```suggestion
          stack.energyFlux += thermalTrans[ke] * m_temp[er][esr][ei] + dThermalTrans_dT[ke] * m_temp[er][esr][ei] * m_temp[er][esr][ei] / 2.0;
```

File: src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst
Comment (jhuang2601) [inline src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst]:
```suggestion
The reference value of the single phase thermal conductivity of the porous medium around the wellbore and its derivative with respect to temperature are defined in the ``SinglePhaseThermalConductivity`` XML block:  
```

File: src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst
Comment (jhuang2601) [inline src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst]: any reference for the analytical solution?

File: src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst
Comment (jhuang2601) [inline src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst]: To avoid unexpected doc build failure, it is better to give a unique name for the python script of each case.

File: src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst
Comment (jhuang2601) [inline src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst]:
```suggestion
This example is an extension of the linear thermal diffusion problem presented in :ref:`AdvancedExamplePureThermalDiffusionWellbore` to model wellbore cooling upon CO2 injection. It uses the thermal single-phase flow solver to model a non-linear thermal diffusion problem around a wellbore where the single phase thermal conductivity of the porous rock depends linearly on the temperature.
```

File: src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentVolumetricHeatCapacity/Example.rst
Comment (jhuang2601) [inline src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentVolumetricHeatCapacity/Example.rst]: Why some difference between GEOS and FDM Non-linear are observed for this cooling case? If the case can be refined to show a better match?

**Review by jhuang2601 — APPROVED** [review]
Thanks for implementing this feature!

**Comment by jhuang2601** [comment]
These failures are expected and rebaseline is required.
```
=======================
Test failures
=======================
3D_10x10x10_thermalCompressible_smoke_01
3D_10x10x10_thermalCompressible_smoke_08
CasedThermoElasticWellbore_ImperfectInterfaces_smoke_01
CasedThermoElasticWellbore_smoke_01
CasedThermoElasticWellbore_smoke_04
fractureMatrixThermalFlow_conforming_smoke_01
fractureMatrixThermalFlow_conforming_smoke_04
fractureMatrixThermalFlow_edfm_smoke_01
fractureMatrixThermalFlow_edfm_smoke_04
thermalCompressible_2d_smoke_01
thermalCompressible_2d_smoke_04
thermalCompressible_temperatureDependentSinglePhaseThermalConductivity_smoke_01
thermalCompressible_temperatureDependentSinglePhaseThermalConductivity_smoke_04
thermalCompressible_temperatureDependentVolumetricHeatCapacity_smoke_01
thermalCompressible_temperatureDependentVolumetricHeatCapacity_smoke_04
ThermoPoroElastic_conforming_smoke_01
ThermoPoroElastic_conforming_smoke_04
ThermoPoroElastic_consolidation_smoke_fim_01
ThermoPoroElastic_consolidation_smoke_fim_02
ThermoPoroElastic_consolidation_smoke_sequential_01
ThermoPoroElastic_consolidation_smoke_sequential_02
ThermoPoroElastic_efem-edfm_verticalFrac_smoke_01
ThermoPoroElastic_efem-edfm_verticalFrac_smoke_04
ThermoPoroElasticWellbore_smoke_01
ThermoPoroElasticWellbore_smoke_04
```

File: src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst
Comment (sytuannguyen) [reply src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst]: Done

File: src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst
Comment (sytuannguyen) [reply src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst]: Done

File: src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst
Comment (sytuannguyen) [reply src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst]: Actually the reference results was obtained by the dinite difference method. This is now corrected.

File: src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst
Comment (sytuannguyen) [reply src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentSinglePhaseThermalConductivity/Example.rst]: Done

File: src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp
Comment (sytuannguyen) [reply src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp]: Done

File: src/coreComponents/physicsSolvers/fluidFlow/ThermalSinglePhaseFVMKernels.hpp
Comment (sytuannguyen) [reply src/coreComponents/physicsSolvers/fluidFlow/ThermalSinglePhaseFVMKernels.hpp:361]: I am not sure about this change

File: src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentVolumetricHeatCapacity/Example.rst
Comment (sytuannguyen) [reply src/docs/sphinx/advancedExamples/validationStudies/wellboreProblems/nonLinearThermalDiffusion_TemperatureDependentVolumetricHeatCapacity/Example.rst]: I have tried different experements but I can not get a ferfect fit. This might be because the reference result is also numercial, and there might be numercial errors from both sides.

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (sytuannguyen) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp]: Done

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (CusiniM) [inline src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp]:
```suggestion
  forAll<ParallelDevicePolicy<>>( temperature.size( 0 ), [=] GEOS_HOST_DEVICE ( localIndex const ei )
  {
  
  });
```

Comment (CusiniM) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp]: and then you need to create views for all those member variables your are using inside the loop coz you cannot use member variables inside that raja loop.

For example:

````c++
auto effectiveConductivity = m_effectiveConductivity.toView();
````

Comment (sytuannguyen) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp]: Done

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (CusiniM) [inline src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp:56]: why is this function taking the porosity as an input? 
I am also not sure I understand the meaning of `RockFluidState`

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (CusiniM) [inline src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp:59]: don't you need the stride dimensions for these? Like this, It will probably fail on the gpus.

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (CusiniM) [inline src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp]:
```suggestion
  R1Tensor const  defaultThermalConductivityComponents = m_defaultThermalConductivityComponents;
  R1Tensor const  thermalConductivityGradientComponents = m_thermalConductivityGradientComponents;
```

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (CusiniM) [inline src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp]:
```suggestion
  R1Tensor const defaultThermalConductivityComponents = m_defaultThermalConductivityComponents;
  R1Tensor const thermalConductivityGradientComponents = m_thermalConductivityGradientComponents;
```

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (CusiniM) [inline src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp:79]: this is not really how things are usually done in GEOS since we usually have these update functions in the kernelWrapper objects.

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (sytuannguyen) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp:56]: That function exists before this PR, so I am not sure about it. I did not touch it except your suggestion about the parallel loop.

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (sytuannguyen) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp:79]: Before this PR, there is an "update" function here, I just replace it by this function. By doing so, I did not intend to change the architecture design of GEOS.

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (sytuannguyen) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp:59]: Actually the build is success for all the platforms including gpus. Also, I have validated the implementation against reference results.

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (sytuannguyen) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp]: Done

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (sytuannguyen) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp]: Done

File: inputFiles/singlePhaseFlow/thermalCompressible_2d_base.xml
Comment (CusiniM) [inline inputFiles/singlePhaseFlow/thermalCompressible_2d_base.xml:14]: why? is this case taking 100 nonlinear iterations?

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (CusiniM) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp:56]: well, can we change it? I don't understand why it would use the porosity. And the name is quite confusing.

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivityBase.hpp
Comment (CusiniM) [inline src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivityBase.hpp:147]: do you need this?

**Review by CusiniM — COMMENTED** [review]
Can we find a better name to replace `ThermalConductivityGradientComponents`?

File: inputFiles/singlePhaseFlow/thermalCompressible_2d_base.xml
Comment (jhuang2601) [reply inputFiles/singlePhaseFlow/thermalCompressible_2d_base.xml:14]: 20 would be sufficient

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (jhuang2601) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp:56]: This function is useless for single phase flow and probably can be removed?

**Comment by jhuang2601** [comment]
> Can we find a better name to replace `ThermalConductivityGradientComponents`?

How about `ThermalConductivityGradient`?

File: inputFiles/singlePhaseFlow/thermalCompressible_2d_base.xml
Comment (sytuannguyen) [reply inputFiles/singlePhaseFlow/thermalCompressible_2d_base.xml:14]: This number is to ensure the convergence of the benchmark case

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (sytuannguyen) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp:56]: This function is not belong to this PR, so I propose to leave it as it is, we can change or remove it if we want but I propose to do that in another PR.

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivityBase.hpp
Comment (sytuannguyen) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivityBase.hpp:147]: Yes, absolutly, adding this term is the core of this implementation for temperature dependent single phase thermal conductivity.

**Comment by sytuannguyen** [comment]
> > Can we find a better name to replace `ThermalConductivityGradientComponents`?
> 
> How about `ThermalConductivityGradient`?

From my point of view, ThermalConductivityGradientComponents is more appropriate because it stands for the components of the thermal gradient vector, also it is consistent to the existing name ThermalConductivityComponents for the components of the thermal conductivity.

**Review by CusiniM — APPROVED** [review]

**Review by paveltomin — APPROVED** [review]

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp
Comment (jhuang2601) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp:56]: Add an issue (#3298) for this discussion and will handle it in the following PR.

**Comment by CusiniM** [comment]
This PR still needs a codeowner's review so I cannot merge it. Let's not put in the merge queue PR's that still need approval because they clog it.

**Comment by sytuannguyen** [comment]
> This PR still needs a codeowner's review so I cannot merge it. Let's not put in the merge queue PR's that still need approval because they clog it.

What does this means? Do you mean this PR need to be validated by all the 10 code owners or just some of you?

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.hpp
Comment (dkachuma) [inline src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.hpp:49]: I would have expected the update function to be implemented. Before, with the constant conductivity model, there was nothing to do. Now that we have a temperature dependency, I think we need to implement this function.

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivityBase.hpp
Comment (dkachuma) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivityBase.hpp:147]: Does it need to be in the base class? Can it be in the `SinglePhaseThermalConductivity` class itself. I can imagine we might later have thermal conductivity models that don't need this.

**Review by dkachuma — APPROVED** [review]

**Comment by CusiniM** [comment]
> > This PR still needs a codeowner's review so I cannot merge it. Let's not put in the merge queue PR's that still need approval because they clog it.
> 
> What does this means? Do you mean this PR need to be validated by all the 10 code owners or just some of you?

Only 1 code owner for each folder that was modified by the PR.

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivityBase.hpp
Comment (CusiniM) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivityBase.hpp:147]: I meant "do you need this getter"?  We have specific ways to grab fields from consitutive models through `getField`

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.hpp
Comment (CusiniM) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.hpp:49]: they seem to be launching the update through this function: https://github.com/GEOS-DEV/GEOS/blob/2647b3176e2beca577f66623e2cf7ce37cbff928/src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.cpp#L62

**Review by corbett5 — APPROVED** [review]

**Comment by jhuang2601** [comment]
Hello @rrsettgast, please take a look on this PR, which is waiting for your approval.

File: src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.hpp
Comment (dkachuma) [reply src/coreComponents/constitutive/thermalConductivity/SinglePhaseThermalConductivity.hpp:49]: I see. It's quite a different pattern.

File: inputFiles/singlePhaseFlow/thermalCompressible_2d_base.xml
Comment (rrsettgast) [reply inputFiles/singlePhaseFlow/thermalCompressible_2d_base.xml:14]: I should converge in 20. if not, then something is wrong.

**Review by rrsettgast — APPROVED** [review]

File: inputFiles/singlePhaseFlow/thermalCompressible_2d_base.xml
Comment (jhuang2601) [reply inputFiles/singlePhaseFlow/thermalCompressible_2d_base.xml:14]: I can confirm that it converges in 20.

**Review by jhuang2601 — APPROVED** [review]

**Pull request merged**
